spec/awful/popup_spec.lua

```diff
@@ -0,0 +1,106 @@
+        set_newindex_miss_handler = function() end,
+    }),
+    {
+        __call = function()
+            return {
+                connect_signal = function() end,
+            }
+        end,
+    }
+)
+package.loaded["wibox.drawable"] = setmetatable({}, {
+    __call = function()
+        return {
+            _inform_visible = function() end,
+            connect_signal = function() end,
+            set_bg = function() end,
+            set_fg = function() end,
+            draw = function() end,
+            set_widget = function() end,
+            get_widget = function() end,
+        }
+    end,
+})
+package.loaded["gears.debug"] =
+    { deprecate = function() end, deprecate_class = function() end }
+
+local popup = require "awful.popup"
+
+describe("awful.popup", function()
+    describe("hide_on_click", function()
+        it(
+            "Constructor without the param should not connect the hide_fct",
+            function()
+                local p = popup { widget = utils.widget_stub() }
+                assert.is_nil(p._signals["button::press"])
+            end
+        )
+
+        it("Constructor parameter should connect the hide_fct", function()
+            local p =
+                popup { widget = utils.widget_stub(), hide_on_click = true }
+            assert.is_true(
+                p._signals["button::press"].strong[p._private.hide_fct]
+            )
+        end)
+
+        it("Setter set_hide_click should connect/disconnect the the hide_fct", function()
+            local p = popup { widget = utils.widget_stub() }
+            assert.is_nil(p._signals["button::press"])
+
+            p:set_hide_on_click(true)
+            assert.is_true(
+                p._signals["button::press"].strong[p._private.hide_fct]
+            )
+
+            p:set_hide_on_click(false)
+            assert.is_nil(
+                p._signals["button::press"].strong[p._private.hide_fct]
+            )
+        end)
+
+        it(
+            "Legacy constructor parameter hide_on_right_click should still work",
+            function()
+                local p = popup {
+                    widget = utils.widget_stub(),
+                    hide_on_right_click = true,
+                }
+                assert.is_true(
+                    p._signals["button::press"].strong[p._private.hide_fct]
+                )
+            end
+        )
+
+        it("Legacy setter set_hide_on_right_click should still work", function()
+            local p = popup { widget = utils.widget_stub() }
+            assert.is_nil(p._signals["button::press"])
+
+            p:set_hide_on_right_click(true)
+            assert.is_true(
+                p._signals["button::press"].strong[p._private.hide_fct]
+            )
+
+            p:set_hide_on_right_click(false)
+            assert.is_nil(
+                p._signals["button::press"].strong[p._private.hide_fct]
+            )
+        end)
+    end)
+end)
```

Comment:
Unit tests shouldn't test private properties and implementation details. They shouldn't know or care about those things.
Another red flag is the fact that you had to noop-stub half the widget system to get this to work. That will never be an accurate test of what's going on in the non-stubbed library.

The integration test in `tests` is the correct way to test functionality like that.

Reply:
Maybe we could make a "global" monkey-patch of `gears.object` and it's CAPI variant to check if a signal is connected. Testing private APIs is bad, but testing at a message passing level is a very good thing. we have plenty of broken signals right now and very little tooling to detect them.

Reply:
Please don't.

If the thing to test for is "if I do A, is signal X emitted?", that's fine. We don't need any monkey-patching for that, just a `spy.new()` passed to `connect_signal` and an `assert.spy().was_called`.

But if the thing to test is "if I do A, does thing X happen?", then the internal signal is an implementation detail that shouldn't be tested. What should be tested is "did X happen?". And that test should be agnostic of whether a signal, function or any other system was called for that to happen, now or in any future revision of the implementation.

The public contract for `hide_on_click` should not include "there is a private function, but you can still access it, and that function is then assigned to another private table based on the specific signal we use internally". There shouldn't be a test breaking because someone decided to make the function actually `local`, renamed it, or because the "when button is clicked" logic changed.

Reply:
Hm, yeah. I didn't feel so good about these unit tests. I'm happy to see some feedback on it. Thank you both of you @sclu1034 and @Elv13.

First, to describe my goal with these unit tests, I want to make sure that setting the parameter, or calling the setter, would actually trigger the appropriate signal connection. So that's a "if I do A, does thing X happen?" as described by @sclu1034. The idea was to make sure the parameter is actually taken into account. (it was mostly in the regard of the recent PR about the `with_shell` parameter blatantly ignored by the implementation.)

> Another red flag is the fact that you had to noop-stub half the widget system to get this to work.

I have to disagree with this one. In a unit test suite, you only load the module that the test targets. In this case, `awful.popup`. However, for the module to work, we need to have some tools from Awesome that just can't be here, like the CAPI. It just doesn't exist when we `require("awful.popup")` with busted.

> Maybe we could make a "global" monkey-patch of gears.object and it's CAPI variant to check if a signal is connected.

> We don't need any monkey-patching for that, just a spy.new() passed to connect_signal and an assert.spy().was_called.

Yup, we could have definition with spies in the `preload.lua` busted helper for this kind of objects.

> The public contract for hide_on_click should not include "there is a private function, but you can still access it, and that function is then assigned to another private table based on the specific signal we use internally". There shouldn't be a test breaking because someone decided to make the function actually local, renamed it, or because the "when button is clicked" logic changed.

This sounds right, but it doesn't give a solution for the "I want to make sure that setting the parameter, or calling the setter, would actually trigger the appropriate signal connection" problem. Hm... This is actually an obvious example of "testing implementation details".

> Maybe we could also make hide_on_click either a boolean or a key/value table of button id as keys and boolean as values? However, that's not required to merge this PR. It's just "umm, that would be nice!".

Making it a key/value table of button id as keys and boolean as values is basically a simplification (would it even be easier?) of the `buttons` property. Sooo :shrug: 

Making `hide_on_click` a boolean property can actually be a good idea. We can make it behave as a property by only adding a getter that does the `self._signals["button::press"].strong[self._private.hide_fct] ~=nil` test.  So the unit test would have to check that after setting the `hide_on_click` value, the getter returns the appropriate boolean value.

I guess this way, we can reconcile both parties.

Reply:
Hm, I guess we could also just test that setting hide_on_click on the constructor will actually call the "set_hide_on_click' method.